apps/admin/backend/src/app.ts

```diff
@@ -558,92 +549,61 @@ function buildApi({
-
-      const ballotStyle = getBallotStyle({
-        ballotStyleId: cvr._ballotStyleId,
-        election,
-      });
+      const { contestId, optionId, cvr } = castVoteRecordData;
       if (cvr._layouts === undefined || cvr._ballotImages === undefined) {
         return []; // The CVR does not have ballot images.
       }
-      if (!ballotStyle) {
-        throw new Error('unexpected types');
-      }
-      const allContestIdsForBallotStyle = getContests({
-        ballotStyle,
-        election,
-      }).map((c) => c.id);
-      const [layouts, ballotImages] = iter(cvr._layouts)
-        .zip(cvr._ballotImages)
-        .toArray()
-        .sort(([a], [b]) => a.metadata.pageNumber - b.metadata.pageNumber)
-        .reduce<[BallotPageLayout[], InlineBallotImage[]]>(
-          ([layoutsAcc, ballotImagesAcc], [layout, ballotImage]) => [
-            [...layoutsAcc, layout],
-            [...ballotImagesAcc, ballotImage],
-          ],
-          [[], []]
+
+      // Identify the page layout, contest layout, and associated image.
+      let pageLayout: Optional<BallotPageLayout>;
+      let contestLayout: Optional<BallotPageContestLayout>;
+      let ballotImage: InlineBallotImage | null | undefined;
+      for (const [index, layout] of cvr._layouts.entries()) {
+        if (layout === null) continue;
+
+        const foundContestLayout = layout.contests.find(
+          (contest) => contest.contestId === contestId
         );
-      let contestIdx = allContestIdsForBallotStyle.indexOf(contestId);
-      let currentLayoutOptionIdx = 0;
-      let currentLayout = layouts[currentLayoutOptionIdx];
-      while (currentLayout && contestIdx >= currentLayout.contests.length) {
-        // move to the next page, past the contests of the current page
-        contestIdx -= currentLayout.contests.length;
-
-        currentLayoutOptionIdx += 1;
-        currentLayout = layouts[currentLayoutOptionIdx];
-        if (!currentLayout) {
-          throw new Error('unexpected types');
+        if (foundContestLayout) {
+          pageLayout = layout;
+          contestLayout = foundContestLayout;
+          ballotImage = cvr._ballotImages[index];
+          continue;
         }
       }
-      currentLayout = layouts[currentLayoutOptionIdx];
-      if (!currentLayout) {
-        throw new Error('unexpected types');
+      if (!pageLayout || !contestLayout) {
+        throw new Error('unable to find a layout for the specified contest');
+      }
+      if (!ballotImage) {
+        throw new Error('no ballot image associated with the ballot layout');
       }
-      const contestLayout =
-        layouts[currentLayoutOptionIdx]?.contests[contestIdx];
 
-      // Options are laid out from the bottom up, so we reverse write-ins to get the correct bounds
-      const writeInOptions = contestLayout?.options
+      // Identify the write-in option layout. Options are laid out from the
+      // bottom up, so we reverse write-ins to get the correct bounds.
```

Comment:
Is this true? Are the write-ins in the layout in reverse order for some reason? That doesn't match with my intuition for the write-in option indices

Reply:
Sounds like you're more on top of this question than I am. I had no idea.